src/main/java/com/amazon/ion/impl/macro/MacroEvaluatorAsIonReader.kt

```diff
@@ -0,0 +1,149 @@
+    override fun stepOut() {
+        evaluator.stepOut()
+        containerStack.pop()
+        // This is essentially a no-op for Lists and SExps
+        currentFieldName = containerStack.peek().currentFieldName
+        currentValueExpression = null // Must call `next()` to get the next value
+        queuedFieldName = null
+        queuedValueExpression = null
+    }
+
+    override fun close() { TODO("Not yet implemented") }
+    override fun <T : Any?> asFacet(facetType: Class<T>?): Nothing = TODO("Not supported")
+    override fun getDepth(): Int = containerStack.size()
+    override fun getSymbolTable(): SymbolTable = TODO("Not implemented in this abstraction")
+
+    override fun getType(): IonType? = currentValueExpression?.type
+
+    override fun getTypeAnnotations(): Array<String>? = currentValueExpression?.annotations?.let { Array(it.size) { i -> it[i].assumeText() } }
+    override fun getTypeAnnotationSymbols(): Array<SymbolToken>? = currentValueExpression?.annotations?.toTypedArray()
+    // TODO: Make this into an iterator that unwraps the SymbolTokens as it goes instead of allocating a new list
+    override fun iterateTypeAnnotations(): MutableIterator<String>? = currentValueExpression?.annotations?.mapTo(mutableListOf()) { it.assumeText() }?.iterator()
+
+    override fun isInStruct(): Boolean = TODO("Not yet implemented")
+
+    override fun getFieldId(): Int = currentFieldName?.value?.sid ?: 0
+    override fun getFieldName(): String? = currentFieldName?.value?.text
+    override fun getFieldNameSymbol(): SymbolToken? = currentFieldName?.value
+
+    override fun isNullValue(): Boolean = currentValueExpression is Expression.NullValue
+    override fun booleanValue(): Boolean = (currentValueExpression as Expression.BoolValue).value
+
+    override fun getIntegerSize(): IntegerSize {
+        // TODO: Make this more efficient, more precise
+        return when (val intExpression = currentValueExpression as Expression.IntValue) {
+            is Expression.LongIntValue -> if (intExpression.value.toInt().toLong() == intExpression.value) {
+                IntegerSize.INT
+            } else {
+                IntegerSize.LONG
+            }
+            is Expression.BigIntValue -> IntegerSize.BIG_INTEGER
+        }
+    }
+
+    /** TODO: Throw on data loss */
+    override fun intValue(): Int = longValue().toInt()
+
+    override fun longValue(): Long = when (val intExpression = currentValueExpression as Expression.IntValue) {
+        is Expression.LongIntValue -> intExpression.value
+        is Expression.BigIntValue -> intExpression.value.longValueExact()
+    }
+
+    override fun bigIntegerValue(): BigInteger = when (val intExpression = currentValueExpression as Expression.IntValue) {
+        is Expression.LongIntValue -> intExpression.value.toBigInteger()
+        is Expression.BigIntValue -> intExpression.value
+    }
+
+    override fun doubleValue(): Double = (currentValueExpression as Expression.FloatValue).value
```

Comment:
🗺️ 

> Error handling is inconsistent with other [IonReader] implementations

This is one specific instance here. If the reader is on e.g. a `StringValue` expression, and you call `doubleValue()`, this will throw a `ClassCastException`. This will get fixed up in subsequent PRs once I have the macro evaluator and I'm able to more easily run tests on this class.